swarm/storage/netstore.go

```diff
@@ -18,318 +18,280 @@ package storage
-
-		return chunk, nil
-	}
-	return fetch(rctx)
-}
+		metrics.GetOrRegisterResettingTimer(fmt.Sprintf("netstore.fetcher.lifetime.%s", fii.CreatedBy), nil).UpdateSince(fii.CreatedAt)
 
-// FetchFunc returns nil if the store contains the given address. Otherwise it returns a wait function,
-// which returns after the chunk is available or the context is done
-func (n *NetStore) FetchFunc(ctx context.Context, ref Address) func(context.Context) error {
-	chunk, fetch, _ := n.get(ctx, chunk.ModeGetRequest, ref)
-	if chunk != nil {
-		return nil
-	}
-	return func(ctx context.Context) error {
-		_, err := fetch(ctx)
-		return err
+		if time.Since(fii.CreatedAt) > 5*time.Second {
+			log.Trace("netstore.put slow chunk delivery", "ref", ch.Address().String())
+		}
+
+		n.fetchers.Delete(ch.Address().String())
 	}
+
+	return exists, nil
 }
 
 // Close chunk store
-func (n *NetStore) Close() (err error) {
-	close(n.closeC)
-
-	wg := sync.WaitGroup{}
-	for _, key := range n.fetchers.Keys() {
-		if f, ok := n.fetchers.Get(key); ok {
-			if fetch, ok := f.(*fetcher); ok {
-				wg.Add(1)
-				go func(fetch *fetcher) {
-					defer wg.Done()
-					fetch.cancel()
-
-					select {
-					case <-fetch.deliveredC:
-					case <-fetch.cancelledC:
-					}
-				}(fetch)
-			}
-		}
-	}
-	wg.Wait()
-
+func (n *NetStore) Close() error {
 	return n.Store.Close()
 }
 
-// get attempts at retrieving the chunk from LocalStore
-// If it is not found then using getOrCreateFetcher:
-//     1. Either there is already a fetcher to retrieve it
-//     2. A new fetcher is created and saved in the fetchers cache
-// From here on, all Get will hit on this fetcher until the chunk is delivered
-// or all fetcher contexts are done.
-// It returns a chunk, a fetcher function and an error
-// If chunk is nil, the returned fetch function needs to be called with a context to return the chunk.
-func (n *NetStore) get(ctx context.Context, mode chunk.ModeGet, ref Address) (Chunk, func(context.Context) (Chunk, error), error) {
-	n.mu.Lock()
-	defer n.mu.Unlock()
-
-	chunk, err := n.Store.Get(ctx, mode, ref)
+// Get retrieves a chunk
+// If it is not found in the LocalStore then it uses RemoteGet to fetch from the network.
+func (n *NetStore) Get(ctx context.Context, mode chunk.ModeGet, req *Request) (Chunk, error) {
+	metrics.GetOrRegisterCounter("netstore.get", nil).Inc(1)
+	start := time.Now()
+
+	ref := req.Addr
+
+	log.Trace("netstore.get", "ref", ref.String())
+
+	ch, err := n.Store.Get(ctx, mode, ref)
 	if err != nil {
-		// TODO: Fix comparison - we should be comparing against leveldb.ErrNotFound, this error should be wrapped.
+		// TODO: fix comparison - we should be comparing against leveldb.ErrNotFound, this error should be wrapped.
 		if err != ErrChunkNotFound && err != leveldb.ErrNotFound {
-			log.Debug("Received error from LocalStore other than ErrNotFound", "err", err)
+			log.Error("got error from LocalStore other than leveldb.ErrNotFound or ErrChunkNotFound", "err", err)
```

Comment:
the error message here can actually be simplified to `localstore get returned error`. the error will not be logged at all if it doesn't pass the check. self explanatory